### PR TITLE
Improve documentation for pipeline parallelism.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49577 Add benchmark for torch.distributed.pipeline.sync.Pipe
* **#48638 Improve documentation for pipeline parallelism.**
* #48630 Cleanup APIs for pipeline parallelism.
* #48470 Test pipeline parallelism works with DDP.

Polishing up some of the docs for the main `Pipe` class and its
`forward` method.

Differential Revision: [D25237705](https://our.internmc.facebook.com/intern/diff/D25237705/)

### CODE REVIEW

File: torch/distributed/pipeline/sync/pipe.py [hunk 1]
```diff
@@ -137,64 +137,65 @@ def _split_module(modules: nn.Sequential) -> Tuple[List[nn.Sequential], List[tor
 
 class Pipe(Module):
     """Wraps an arbitrary :class:`nn.Sequential <torch.nn.Sequential>` module
-    to train on Pipe_. If the module requires lots of memory, Pipe will be
-    very efficient.
-    ::
+    to train on using synchronous pipeline parallelism. If the module requires lots of memory and doesn't fit on a single GPU, pipeline parallelism is a useful technique to employ for training.
```
Comment (mrshenli) [inline torch/distributed/pipeline/sync/pipe.py:140]: nit: shall we break long lines in this file into shorter ones?

File: torch/distributed/pipeline/sync/pipe.py [hunk 2]
```diff
@@ -137,64 +137,65 @@ def _split_module(modules: nn.Sequential) -> Tuple[List[nn.Sequential], List[tor
...
-
-    Keyword Args:
         chunks (int):
             number of micro-batches (default: ``1``)
         checkpoint (str):
             when to enable checkpointing, one of ``'always'``,
-            ``'except_last'``, or ``'never'`` (default: ``'except_last'``)
+            ``'except_last'``, or ``'never'`` (default: ``'except_last'``). ``'never'`` disables checkpointing completely, ``'except_last'`` enables checkpointing for all micro-batches except the last one and ``'always'`` enables checkpointing for all micro-batches.
         deferred_batch_norm (bool):
             whether to use deferred BatchNorm moving statistics (default:
```
Comment (mrshenli) [inline torch/distributed/pipeline/sync/pipe.py:165]: double quote `BatchNorm`?

File: torch/distributed/pipeline/sync/pipe.py [hunk 3]
```diff
@@ -137,64 +137,65 @@ def _split_module(modules: nn.Sequential) -> Tuple[List[nn.Sequential], List[tor
...
+            sequential module to be parallelized using pipelining. Each module
             in the sequence has to have all of its parameters on a single
             device. Each module in the sequence has to either be an nn.Module
-            or ``nn.Sequential`` (to combine multiple sequential modules on a single
+            or :class:`nn.Sequential <torch.nn.Sequential>` (to combine multiple sequential modules on a single
             device)
-
-    Keyword Args:
         chunks (int):
             number of micro-batches (default: ``1``)
```
Comment (mrshenli) [inline torch/distributed/pipeline/sync/pipe.py:160]: This is assuming the first dimension is the batch dimension? Do we want to mention what if `chunks` > the samples in a batch?

File: torch/distributed/pipeline/sync/pipe.py [hunk 4]
```diff
@@ -137,64 +137,65 @@ def _split_module(modules: nn.Sequential) -> Tuple[List[nn.Sequential], List[tor
...
+        >>> input = torch.rand(16, 16).cuda(0)
+        >>> output_rref = model(input)
 
-    #: The checkpoint mode to determine when to enable checkpointing. It is one
-    #: of ``'always'``, ``'except_last'``, or ``'never'``.
-    checkpoint: str = "except_last"
+    .. note::
+        You can wrap a :class:`Pipe` model with
+        :class:`torch.nn.parallel.DistributedDataParallel` only when the
+        checkpoint parameter of :class:`Pipe` is ``'never'``.
```
Comment (mrshenli) [inline torch/distributed/pipeline/sync/pipe.py:187]: are we going to check this?

File: torch/distributed/pipeline/sync/pipe.py [hunk 5]
```diff
@@ -306,21 +307,24 @@ def _ensure_copy_streams(self) -> List[List[AbstractStream]]:
...
-        """:class:`Pipe` is a fairly transparent module wrapper. It doesn't
+    def forward(self, input) -> RRef:  # type: ignore
+        """
+        Processes a single input mini-batch through the pipe and returns an
+        :class:`~torch.distributed.rpc.RRef` pointing to the output.
+        :class:`Pipe` is a fairly transparent module wrapper. It doesn't
         modify the input and output signature of the underlying module. But
         there's type restriction. Input and output have to be a
         :class:`~torch.Tensor` or a sequence of tensors. This restriction is
         applied at partition boundaries too.
```
Comment (mrshenli) [inline torch/distributed/pipeline/sync/pipe.py:318]: Any reason for this limitation? Are we going to drop this in the future?

**Review by mrshenli — APPROVED** [review]
LGTM! Left some minor comments.

File: torch/distributed/pipeline/sync/pipe.py (hunk 4 above)
Comment (pritamdamania87) [reply torch/distributed/pipeline/sync/pipe.py:195]: No we are not checking this, this is just a note to the user that other modes will not work with DDP.

File: torch/distributed/pipeline/sync/pipe.py (hunk 5 above)
Comment (pritamdamania87) [reply torch/distributed/pipeline/sync/pipe.py:334]: Its tricky to split an arbitrary python object into multiple microbatches, that is the primary motivation of this limitation.

File: torch/distributed/pipeline/sync/pipe.py (hunk 3 above)
Comment (pritamdamania87) [reply torch/distributed/pipeline/sync/pipe.py:163]: This clarification probably belongs in the forward method, I've updated those docs to reflect this.
